src/pudl/extract/ferceqr.py

```diff
@@ -0,0 +1,224 @@
+
+    @property
+    def base_path(self) -> UPath:
+        """Return UPath pointing to archive base path."""
+        return UPath(self.archive)
+
+
+@contextmanager
+def _get_csv(base_path: UPath, year_quarter: str) -> zipfile.ZipFile:
+    """Download CSV to a tempmorary directory to avoid reading into memory."""
+    year, quarter = year_quarter.split("q")
+    zip_name = f"ferceqr-{year}q{quarter}.zip"
+    remote_path = base_path / zip_name
+
+    # Create temp directory to download zip to
+    with (
+        tempfile.TemporaryDirectory() as tmp_dir,
+    ):
+        # Download file to local path
+        local_path = Path(tmp_dir) / zip_name
+        local_path.write_bytes(remote_path.read_bytes())
+        # Yield open zipfile
+        with zipfile.ZipFile(local_path) as zf:
+            yield zf
+
+
+def _clean_csv_name(csv_path: Path) -> Path:
+    """Standardize zip file names to avoid errors when opening."""
+    new_path = csv_path
+    if "'" in csv_path.name:
+        new_path = csv_path.rename(csv_path.parent / csv_path.name.replace("'", ""))
+    if '"' in csv_path.name:
+        new_path = csv_path.rename(csv_path.parent / csv_path.name.replace('"', ""))
+    return new_path
+
+
+def _get_table_name(table_type: str, year_quarter: str) -> str:
+    if table_type != "indexPub":
+        return f"raw_ferceqr__{table_type}_{year_quarter}"
+    return f"raw_ferceqr__index_pub_{year_quarter}"
+
+
+def _extract_ident(
+    ident_csv: str,
+    year_quarter: str,
+    duckdb_connection: DuckDBPyConnection,
+) -> str:
+    """Extract data from ident csv, write to parquet, and return CID from table.
+
+    This table is always extracted first so we can pull the CID from it and include
+    a CID column in all other tables.
+    """
+    # Use duckdb to read CSV and write as parquet
+    csv_rel = duckdb_connection.read_csv(
+        ident_csv, all_varchar=True, store_rejects=True, ignore_errors=True
+    )
+    (cid,) = csv_rel.select("company_identifier").limit(1).fetchone()
+    persist_table_as_parquet(
+        csv_rel.select(f"*, '{year_quarter}' AS year_quarter"),
+        table_name=_get_table_name("ident", year_quarter),
+        partitions={"cid": cid},
+    )
+    return cid
+
+
+def _extract_other_table(
+    table_type: str,
+    csv_path: str,
+    year_quarter: str,
+    cid: str,
+    duckdb_connection: DuckDBPyConnection,
+):
+    """Extract data from ident csv and write to parquet, returning CID from table."""
+    # Use duckdb to read CSV and write as parquet
+    persist_table_as_parquet(
+        duckdb_connection.read_csv(
+            csv_path, all_varchar=True, store_rejects=True, ignore_errors=True
+        ).select(f"*, '{year_quarter}' AS year_quarter, '{cid}' as company_identifier"),
+        table_name=_get_table_name(table_type, year_quarter),
+        partitions={"cid": cid},
+    )
+
+
+def _csvs_to_parquet(
+    csv_path: Path,
+    year_quarter: str,
+    duckdb_connection: DuckDBPyConnection,
```

Comment:
What is the motivation for using a persistent DuckDB connection, when we aren't even connecting to a database, and are just reading CSVs and writing Parquet files? Is it just resource intensive to create and destroy repeatedly? Is it related to storing the error logs? Can you explain in the docstring?

Reply:
We have to keep the connection open so because duckdb collects the extraction errors as we go and then writes them to errors table at the end. I've added a note about this in the docstring.

Reply:
Is this an in-memory DuckDB database by default?  If we were to generate a massive number of errors for some reason could we run out of memory?  Would it be safer and also workable to put it on disk?